src/operations.c

```diff
@@ -25,31 +25,45 @@
 #include "baton.h"
 #include "operations.h"
 
+// Mutex protecting the connection and the run_timeout_thread flag
+pthread_mutex_t conn_mutex = PTHREAD_MUTEX_INITIALIZER;
 // The connection used by iterate_json and connection_timeout
 rcComm_t *connection;
-// Mutex protecting the connection
-pthread_mutex_t conn_mutex = PTHREAD_MUTEX_INITIALIZER;
-
-// While true, the client will refresh the connection
-int connection_timeout_flag = 1;
+// While true, the client will continue to run the timeout thread
+int run_timeout_thread = 1;
+// Condition variable to exit the timeout thread when work is complete
+pthread_cond_t watchdog_cond = PTHREAD_COND_INITIALIZER;
 
 // Refresh the connection every timeout seconds
 void *connection_timeout(void *timeout) {
     int tsec = *((int *) timeout);
 
-    struct timespec t;
-    t.tv_sec = tsec;
-    t.tv_nsec = 0;
+    struct timespec abs_timeout;
 
-    while (connection_timeout_flag) {
-        nanosleep(&t, NULL);
+    while (1) {
+        pthread_mutex_unlock(&conn_mutex);
```

Comment:
There's something wrong with the locking here - lots of unlocks and no locks.

The mutex needs to be [locked before calling `pthread_cond_timedwait()`](https://linux.die.net/man/3/pthread_cond_timedwait).  You can pretty much lock before the main loop, and unlock after it.  The lock will be dropped automatically while waiting inside `pthread_cond_timedwait()`.

Reply:
Oops - I typo'd bigtime! That `unlock` under `while(1)` is meant to be a lock ... point taken though - I could just surround the lot

Reply:
And I had a spurious unlock in the middle for some reason. It reads better now and seems to be behaving when I have a 10 second timeout while feeding it jobs every 2 seconds - the connection gets closed/reopened as I'd hope.

Reply:
It's looking better now, although you could still move the locking to the outside of the `while (1) { ... }` loop.  Currently every time you go around you do `pthread_mutex_unlock()` followed immediately by `pthread_mutex_lock()` which is a bit pointless.  If you also move the unlock just before the `break` outside, then you can move the check on `run_timeout_thread` to the loop condition and you get:
```
pthread_mutex_lock(&conn_mutex);
while (run_timeout_thread) { ... }
pthread_mutex_unlock(&conn_mutex);
```
which is, I think, a bit cleaner.